http/cookies.lua

```diff
@@ -0,0 +1,325 @@
+local http_patts = require "lpeg_patterns.http"
+local util = require "http.util"
+local psl = require "psl"
+
+local function parse_set_cookie(text_cookie, host, path, time)
+	assert(time, "missing time value for cookie parsing")
+	local key, value, matched_cookie = assert(http_patts.Set_Cookie:match(text_cookie, 1))
+	local cookie = {
+		creation = time;
+		last_access = time;
+		persistent = not not (matched_cookie.expires or matched_cookie["max-age"]);
+		domain = matched_cookie.domain or host;
+		path = matched_cookie.path or path;
+		secure = matched_cookie.secure or false;
+		http_only = matched_cookie.httponly or false;
+		key = key;
+		value = value;
+		host_only = not not matched_cookie.domain;
+		same_site = matched_cookie.same_site;
+	}
+	if matched_cookie["max-age"] then
+		cookie.expires = time + tonumber(matched_cookie["max-age"])
+	else -- luacheck: ignore
+		-- ::TODO:: make use of `expires` cookie value
+	end
+	return cookie
+end
+
+local function bake_cookie(data)
+	assert(data.key, "cookie must contain a `key` field")
+	assert(type(data.key) == "string", "`key` field for cookie must be string")
+	assert(data.value, "cookie must contain a `value` field")
+	assert(type(data.value) == "string", "`value` field for cookie must be string")
+	local cookie = {data.key .. "=" .. data.value}
+	if data.expires then
+		cookie[#cookie + 1] = "; Expires=" .. util.imf_date(data.expires)
+	end
+	if data.max_age then
+		cookie[#cookie + 1] = "; Max-Age=" .. string.format("%d", data.max_age)
+	end
+	if data.domain then
+		cookie[#cookie + 1] = "; Domain=" .. data.domain
+	end
+	if data.path then
+		cookie[#cookie + 1] = "; Path=" .. util.encodeURI(data.path)
+	end
+	if data.secure then
+		cookie[#cookie + 1] = "; Secure"
+	end
+	if data.http_only then
+		cookie[#cookie + 1] = "; HttpOnly"
+	end
+	-- This component is not a part of the RFC 6265 specification for the
+	-- headers, but is instead from a draft of another RFC that builds on the
+	-- original one.
+	-- https://tools.ietf.org/html/draft-ietf-httpbis-cookie-same-site-00#section-4.1
+	if data.same_site then
+		local v
+		if data.same_site:lower() == "strict" then
+			v = "; SameSite=Strict"
+		elseif data.same_site == "Lax" then
+			v = "; SameSite=Lax"
+		else
+			error('invalid value for same_site, expected "Strict" or "Lax"')
+		end
+		cookie[#cookie + 1] = v
+	end
+	return table.concat(cookie)
+end
+
+local function iterate_cookies(cookie)
+	return pairs(assert(http_patts.Cookie:match(cookie, 1)))
+end
+
+local function parse_cookies(cookie)
+	local cookies = {}
+	for k, v in iterate_cookies(cookie) do
```

Comment:
I'm not copying the table. The `iterate_cookies` function returns a `pairs()` iterator after asserting that the cookie does parse correctly (which I added an error message to, after reading this, in case parsing fails)